bd-session-replay/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+          if self.reporting_interval.is_some() && self.is_periodic_reporting_enabled => {
+          log::debug!("session replay recorder capturing wireframe");
+          // We capture a wireframe once per 3s (by default) so backlogging shouldn't be a problem
+          // here (since taking a screenshot takes not more than ~tens of ms).
+          // TODO(Augustyniak): Consider changing the implementation so that we do not ask platform layer
+          // for more wireframes until we receive the previous one.
+          self.target.capture_wireframe();
+        },
+        _ = self.reporting_interval_rate_flag.changed() => {
+          self.reporting_interval = Some(
+            Self::create_interval(
+              self.reporting_interval_rate_flag.read_mark_update(),
+              false
+            )
+          );
+        },
+        _ = self.take_screenshot_rx.recv() => {
+          if !self.is_take_screenshots_enabled {
+            log::debug!("session replay recorder ignored screenshot: taking screenshots is disabled");
+            continue;
+          }
+
+          if !self.is_ready_to_take_screenshot {
+            log::debug!("session replay recorder ignored screenshot: not ready to take screenshot");
+            continue;
+          }
+
+          log::debug!("session replay recorder taking screenshot");
+
+          self.target.take_screenshot();
+
+          // Reset the flag to indicate that the recorder is not ready to take a screenshot until
+          // `ScreenshotLogInterceptor` intercepts a log containing a screenshot that was just requested.
+          self.is_ready_to_take_screenshot = false;
+        },
+        _ = async { self.next_screenshot_rx.recv().await } => {
+          log::debug!("session replay recorder received screenshot");
+          self.is_ready_to_take_screenshot = true;
+        },
+        _ = self.is_periodic_reporting_enabled_flag.changed() => {
+          self.is_periodic_reporting_enabled
+            = self.is_periodic_reporting_enabled_flag.read_mark_update();
+        },
+        _ = self.is_take_screenshots_enabled_flag.changed() => {
+          self.is_take_screenshots_enabled
+            = self.is_take_screenshots_enabled_flag.read_mark_update();
+        },
+        () = &mut local_shutdown => {
+          return;
+        },
+      }
+    }
+  }
+}
+
+//
+// TakeScreenshotHandler
+//
+
+#[derive(Clone)]
+pub struct TakeScreenshotHandler {
+  take_screenshot_tx: Sender<()>,
+}
+
+impl TakeScreenshotHandler {
+  pub fn take_screenshot(&self) {
+    // We use a channel with a capacity of one and employ `try_send` to avoid waiting for space in
+    // the channel before sending a new signal. This is especially important as the method is
+    // expected to be called on logging hot path.
+    // The channel may become full if "take screenshot" requests arrive faster than the platform
+    // layer can process them. In such cases, new requests are ignored.
```

Comment:
Ah ok yeah this makes sense. Forget this is going from the logging code to the recorder.